### ISSUE TITLE
Adding URN and changing Channel in same flow fails

### ISSUE THREAD

**Issue by tybritten** [rapidpro/mailroom#145]
TL;DR-
Condition- Create a flow that both adds a new URN and also tries to change that new channel to the preferred one.
Results- new URN for the new channel is added, but the old channel remains the default, no error logged
Expected- new channel should begin being used immediately after changing the priority in the flow.

Longer:

A flow that:
1) Updates Contact with a new URN (telegram)
2) sends a message on the old channel (SMS) confirming
3) Sets the channel to telegram
4) sends a new message (telegram)

In this case, both messages (2 and 4) go over SMS, and SMS remains the default channel.

From what I can tell in the mailroom code, the adding of the urn (contact_urn_changed) is batch processed at the end of the session so when step 3 happens as a channel event, it tries to update the urn priority but it silently fails because the new URN hasn't been added yet because the session is still ongoing.
two things:
 shouldn't there be a specific channel event type for changechannel (which would still call UpdatePreferredURN) but would log an error instead of ignoring a noop? Since we're specifically asking for a channel priority change, the inability to change it is a failure
shouldn't any major contact changes (like adding a URN) be made synchronously in the flow so it's available to other ops?

**Comment by nicpottier** [comment]
When you say send a message, that is a reply action, not a broadcast ya?

I'm a bit surprised that both messages are going to SMS, I would have expected maybe the Telegram message to fail if this was a problem of that not existing.

Picking what channel to send on is a concern of GoFlow, so I'm guessing the bug here is likely somewhere in there. Adding a URN does not make it the highest priority (setting the channel does), but URNs created within GoFlow are slightly different than those passed in (they don't have IDs) so I'm guessing that is one place where things are going sideways. I'm pretty sure GoFlow should be setting what channel/URN should be used when sending a message, not Mailroom, but as you said there may be a bug there with brand new URNs. @rowanseymour any thoughts there?

In general Mailroom tries very hard not to do interact with the DB with queries that affect only a single row, that just doesn't work when someone adds a million people to a flow. Inserting URNs may need to be an exception to that though as there is the need for that foreign key when creating the message, so that is probably something else to change. Likely acceptable as that isn't super common path though would be great to avoid, will have to think about that.

Actions and events aren't necessarily 1-1. Events aren't created if the action had no effect for example (such as adding a contact to a group they were already in). And in this case, setting the channel as an action really is only changing URN priorities / ordering, so it is more naturally represented that way as a URN updated event. So on that front I think we are fine.

**Comment by tybritten** [comment]
Yes, this is a reply, not a broadcast. We have another flow that's just the channel switch (for people who already have telegram setup) and that one works as expected. 

Hmm yeah I'm still learning how mailroom/courier/goflow work (i'm not super experienced in golang as you can tell from my PRs haha). 
That makes a lot of sense on what may be happening. Maybe there's a cleaner way to deal with only this type of circumstances. We've tested splitting the flows after step 2 (another flow to just switch channel and send final message) but that doesn't work as it seems they're still all part of the same event/session so the same problem applies.

**Comment by nicpottier** [comment]
Ya, to be clear, this is a bug, we will fix it, just not positive on the root cause just yet, but this is certainly a supported usecase.

And really appreciate the good bug reports, you may be new to golang but you have been very useful in paring down the repro steps and even the hunches on what is happening have been great, so much appreciated.

**Comment by tybritten** [comment]
This is a good learning experience for me to understand the nuts and bolts of how RP works under the covers

**Comment by rowanseymour** [comment]
Think the problem is https://github.com/nyaruka/goflow/blob/master/flows/contact.go#L329. We currently only consider tel URNs to be re-assignable across channels. So seems like we need to open that up for other URN types that aren't tied to a single channel.

Also.. this feels like another case where `set_contact_scheme` action would be preferable to a `set_contact_channel` action.

**Comment by nicpottier** [comment]
Ya, it's a bit of a weird one, because Telegram URNs ARE unique per channel, but in this case they are being tied from another contact, so it is actually ok. Not sure how we could lock down that behavior, may just be something we accept as user's possibly creating connections that are invalid.

IE, this scenario won't work:
1. Add a telegram URN to a contact from Telegram channel 1
2. Set preferred channel to Telegram channel 2

Scheme setting would help in this case, though scheme setting doesn't address the multiple shortcodes on one account case.

**Comment by nicpottier** [comment]
(note that we allow people to add random URNs that are invalid already so that may not be any worse than we already are)

**Comment by tybritten** [comment]
so this isn't moving betwen two telegram channels- its moving from SMS to telegram.

**Comment by nicpottier** [comment]
Yep, yep, understood, we are just talking through ramifications.

**Comment by tybritten** [comment]
looking at goflow, i'm not sure why it doesn't work- if AddURN adds the telegram urn to the contact struct, and then UpdatePreferredChannel should still work because the flow is still working against that contact object, yes?

**Comment by rowanseymour** [comment]
With the exception of tel URNs, `UpdatePreferredChannel` just looks for a URN on the contact with that channel and moves it to the front of the contact's list of URNs. Effectively meaning that URN and channel will be used for the next send. However in this case, the new telegram URN doesn't have any channel so doesn't get moved.

**Comment by tybritten** [comment]
if a urn doesn't have a channel shouldn't we set it then when we create it? or when we attempt to `UpdatePreferredChannel` and the channel doesn't exist?

**Comment by tybritten** [comment]
something like 
```
		for _, urn := range c.urns {
			// tel URNs can be re-assigned, other URN schemes are considered channel specific
			if urn.URN().Scheme() == urns.TelScheme && channel.SupportsScheme(urns.TelScheme) {
				urn.SetChannel(channel)
			}

			// if channel is missing (new urn), add it
			if urn.Channel() == nil {
				newchannel := c.assets.Channels().GetForURN(urn, assets.ChannelRoleSend)
				urn.SetChannel(newchannel)
			}

			// move any URNs with this channel to the front of the list
			if urn.Channel() == channel {
				priorityURNs = append(priorityURNs, urn)
			} else {
				otherURNs = append(otherURNs, urn)
			}
		}
```

**Comment by nicpottier** [comment]
At creation time no we shouldn't. When we explicitly set a preferred then probably for all schemes that that channel supports.

Setting a channel is setting URN affinity and that has important consequences in how messages are sent and what channel is used when sent. We don't want to set the channel for a URN unless it is either an explicit action in a flow or due to a contact messaging us.

**Comment by rowanseymour** [comment]
Yeah so I think it's ok for us to change `UpdatePreferredChannel` to set channel if it isn't already set on a URN, provided that channel supports that URN's scheme. There's still the possibility of setting the URN's channel to something that can't send to that URN.. but 🤷‍♂ 

@nicpottier?

**Comment by tybritten** [comment]
by looking the channel up by URN doesn't that generally mean that it is compatible with that urn?

**Comment by nicpottier** [comment]
> There's still the possibility of setting the URN's channel to something that can't send to that URN.. but 🤷‍♂

Yep, think we have to do that, I am made more comfortable about that due to the Kannel case which is basically the same anyways.

> by looking the channel up by URN doesn't that generally mean that it is compatible with that urn?

The answer to that is "maybe" or perhaps more optimistically "probably". Telegram ids for example can only be sent to AFTER they have communicated with your bot, so you can get yourself into trouble there. Same thing with Facebook channels. Even for telephone channels, the channel may be a direct connection to a carrier SMSC via Kannel and that SMSC may only accept messages to numbers that are subscribers of theirs for example.

But I think in the end this just ends up being a user / flow building error in the end.

**Comment by tybritten** [comment]
what about something like this?
```
			if urn.Channel() == nil {
				if channel.SupportsScheme(urn.URN().Scheme()) {
					urn.SetChannel(channel)
				}
			}
```

**Comment by tybritten** [comment]
or does it make more sense to allow the channel to be selected when building the flow on the addurn part during the flow building? a checkbox that says 'assign urn to channel' and a dropdown to pick a channel?

**Comment by rowanseymour** [comment]
I'm leaning toward letting the user pick the channel on the Add URN action. Some of these URN types don't make so much sense without a channel

**Comment by nicpottier** [comment]
Really? Why? That seems really edge-case and not where we should start. In 99% of cases people will have one Telegram channel / Facebook whatever channel connected so there really isn't any ambiguity there. Again, in truth this is no different than the tel case, we shouldn't act differently than we do there for cases that are unlikely to come up.

**Comment by nicpottier** [comment]
Also, we know channel_uuids are fragile in flow definitions, so would be nice to avoid introducing more actions that depend on them.

**Comment by rowanseymour** [comment]
Totally agree that tel shouldn't be treated differently. Maybe I'm misunderstanding the scenario here but to me it seems odd to allow the user to add a URN without a channel, knowing that the URN maybe un-sendable, unless there's a subsequent `set_contact_channel` (in the case of users having more than one channel of a non-tel type). It seems preferable to do that in one action so the URN isn't left in an ambiguous state.

I also assumed that we were working toward replacing `set_contact_channel` with `set_contact_scheme`, no?

And adding a channel dropdown to Add URN seems a fairly trivial addition that solves this for all scenarios (tel and non-tel)

**Comment by rowanseymour** [comment]
Well perhaps such approaches aren't mutually exclusive anyway, and I already made a PR for the first approach https://github.com/nyaruka/goflow/pull/720

**Comment by nicpottier** [comment]
I guess I was thinking of it opposite, that we allow deriving channels for schemes other than `tel` since in reality they are no different. IE, the URN isn't unsendable (necessarily) without a channel, it just might be. (same as for a `tel` channel)

To translate this all to `tel` world, the scenario Tyler is describing is basically:
 1. Have an account with two Kannel channels connected to SMSCs (MTN and Tigo)
 2. Add a new (only) phone number (no channel associated with it)
 3. If they try to send and the most recent channel is MTN and it's a Tigo number, it will fail
 4. Set preferred channel to Tigo, send a message, it will succeed

It is no different than the Telegram channel.

**Comment by nicpottier** [comment]
I don't want to create another `channel_uuid` reference in goflow, that's something we should be moving away from, not adding more of.

**Comment by tybritten** [comment]
fixed in https://github.com/nyaruka/goflow/pull/720

**Comment by nicpottier** [comment]
I don't think this is actually fixed, the goflow side is only half the equation no? Did you try it?

**Comment by tybritten** [comment]
Yes, and it's broken haha.

**Comment by tybritten** [comment]
i'm trying to figure out why though. If goflow reorders the URNs properly, why cant `UpdatePreferredURN` handle it as is?

**Comment by nicpottier** [comment]
I'm guessing the `if` in that sentence is the rub. :) 

I have some things due early this week but will look at this after those are taken care of.

**Comment by nicpottier** [comment]
(or could be that the create message part of the code blows up when the event itself has a URN that doesn't have an id in it)

**Comment by tybritten** [comment]
I'll probably poke around at it some more this afternoon. The odd thing wasn't that it failed to reorder the urns and the behavior stayed the same. The initial SMS message before the channel change failed too.

**Comment by nicpottier** [comment]
Hrmm, that wouldn't be surprising if the transaction as a whole is failing. (perhaps due to inserting a `NULL` id)

**Comment by tybritten** [comment]
yup, digging through the logs- `attempt to create messaging message with 0 id URN`

**Comment by nicpottier** [comment]
Ya, so that's here:
  https://github.com/nyaruka/mailroom/blob/master/hooks/msg_created.go#L195

So I think in that block instead of an error we could do a URN lookup (with no backdown to a create) using the TX passed in and that should work. If we don't get a URN back then that should remain an error. (broadcasts to brand new URNs aren't this event and go through a different path that handles brand new URNs)

So just an addition of `URNForURN` and handling of the error / nil case, then updating the Msg with the result should do it.

Apologies, still have other stuff on my plate until Thursday but I think the above should be the roadmap to fixing this. :)

**Comment by nicpottier** [comment]
Actually the above may need to live in `models.NewOutgoingMsg` since the GoFlow Msg is immutable.

**Comment by tybritten** [comment]
interesting that `NewOutgoingMsg` doesn't fail on setURN if its nil

**Comment by nicpottier** [comment]
Not all messages have URNs. Specifically messages collected by the offline surveyor app have no URN, only a contact.

**Comment by nicpottier** [comment]
Just FYI so we don't walk on each other addressing this now.

**Comment by nicpottier** [comment]
Should be addressed in v2.0.11, let us know if you see otherwise.

**Comment by chris-erickson** [comment]
Looks like we're all set now!

**Comment by nicpottier** [comment]
Cool. Note that v2.0.12 fixed a db connection leak you may want to ship sooner rather than later.

**Comment by chris-erickson** [comment]
yep we shipped `v2.0.13`
